storage/innobase/include/trx0purge.h

```diff
@@ -55,80 +55,78 @@ Run a purge batch.
 @return number of undo log pages handled in the batch */
 ulint trx_purge(ulint n_tasks, ulint history_size);
 
-/** Rollback segements from a given transaction with trx-no
-scheduled for purge. */
-class TrxUndoRsegs {
-private:
-	typedef std::vector<trx_rseg_t*, ut_allocator<trx_rseg_t*> >
-		trx_rsegs_t;
-public:
-	typedef trx_rsegs_t::iterator iterator;
-	typedef trx_rsegs_t::const_iterator const_iterator;
-
-	TrxUndoRsegs() = default;
-
-	/** Constructor */
-	TrxUndoRsegs(trx_rseg_t& rseg)
-		: trx_no(rseg.last_trx_no()), m_rsegs(1, &rseg) {}
-	/** Constructor */
-	TrxUndoRsegs(trx_id_t trx_no, trx_rseg_t& rseg)
-		: trx_no(trx_no), m_rsegs(1, &rseg) {}
-
-	bool operator!=(const TrxUndoRsegs& other) const
-	{ return trx_no != other.trx_no; }
-	bool empty() const { return m_rsegs.empty(); }
-	void erase(iterator& it) { m_rsegs.erase(it); }
-	iterator begin() { return(m_rsegs.begin()); }
-	iterator end() { return(m_rsegs.end()); }
-	const_iterator begin() const { return m_rsegs.begin(); }
-	const_iterator end() const { return m_rsegs.end(); }
-
-	/** Compare two TrxUndoRsegs based on trx_no.
-	@param elem1 first element to compare
-	@param elem2 second element to compare
-	@return true if elem1 > elem2 else false.*/
-	bool operator()(const TrxUndoRsegs& lhs, const TrxUndoRsegs& rhs)
-	{
-		return(lhs.trx_no > rhs.trx_no);
-	}
-
-	/** Copy of trx_rseg_t::last_trx_no() */
-	trx_id_t trx_no= 0;
-private:
-	/** Rollback segments of a transaction, scheduled for purge. */
-	trx_rsegs_t m_rsegs{};
-};
-
-typedef std::priority_queue<
-	TrxUndoRsegs,
-	std::vector<TrxUndoRsegs, ut_allocator<TrxUndoRsegs> >,
-	TrxUndoRsegs>	purge_pq_t;
-
-/** Chooses the rollback segment with the oldest committed transaction */
-struct TrxUndoRsegsIterator {
-	/** Constructor */
-	TrxUndoRsegsIterator();
-	/** Sets the next rseg to purge in purge_sys.
-	Executed in the purge coordinator thread.
-	@retval false when nothing is to be purged
-	@retval true  when purge_sys.rseg->latch was locked */
-	inline bool set_next();
-
-private:
-	// Disable copying
-	TrxUndoRsegsIterator(const TrxUndoRsegsIterator&);
-	TrxUndoRsegsIterator& operator=(const TrxUndoRsegsIterator&);
-
-	/** The current element to process */
-	TrxUndoRsegs			m_rsegs;
-	/** Track the current element in m_rsegs */
-	TrxUndoRsegs::const_iterator	m_iter;
-};
-
 /** The control structure used in the purge operation */
 class purge_sys_t
 {
-  friend TrxUndoRsegsIterator;
+  /** Min-heap based priority queue over fixed size array */
+  class purge_queue_t
```

Comment:
According to https://stackoverflow.com/questions/231760/what-does-a-type-followed-by-t-underscore-t-represent the `_t` suffix is reserved by POSIX for the implementation. We have been hit by that in the past; the
```c
#define lock_t ib_lock_t
```
is there to allow the code to be compiled on some system. I would not use the `_t` suffix in new data types.